### ISSUE TITLE
Editor scene switching has 40 second delays in some cases.

### ISSUE THREAD

**Issue by avencherus** [godotengine/godot#17400]
The following is happening for me using 2.1.4 on Windows, and on multiple machines.

I have a level scene, which has about 10 scenes in it I named segments.  Each segment scene has about 6 to 10 smaller scenes called platforms.

This scene loads fine in the editor, but whenever I click a tab to switch away from that scene to another one in the editor, it will lock the editor for about 38-40 seconds before switching.

I've cut it up, removed tool scripts, redid the whole thing, and it just seems that the more items that exist at the wider end of the scene tree the longer these delays occur.  As there are more platform scenes the greater the delay, but the platforms are only static bodies with a texture (using .atex), collider, and a couple position 2D nodes.

Is there a known bug, a work around, or something I should check?

It seems like maybe it is stuck doing something and then yielding to a time out.

As small as it seems at first, it is actually now adding up to a lot of extra time in development. :(

Any suggestions would be appreciated.

-Thanks

**Comment by girng** [comment]
Example project please ^_^

**Comment by reduz** [comment]
yeah, this happens because the scenes are packed before they are switched.
It would be interesting to know what is taking so long, though.
If you can make an example project with only what takes time for the
switching, it should be easy to optimize it.

**Comment by avencherus** [comment]
@reduz @girng

I'll see if I can reproduce it in a smaller project.  The current one has gotten worse as we added to it, it will hang for about 2-3 minutes, and if you exit while on it, it will crash the editor.

In the meanwhile some updates that might be helpful:  

The culprit from what I can tell so far is that I have a startup scene (assigned to F5 play), it will instance this large level scene in question, and add it as a child.

It is after this point that it seems like that scene is staying in the editor memory, so maybe I'm supposed to free it when I quit the program or can a leak in a game object linger about in editor memory in some unexpected way?  I'm suspecting not, because this persists even after restarting the editor, something is saved.

So what I found is that when I copy the level, this copy be tabbed out of normally, no problems.  If I then switch the path to load this copied scene.  After it is played once in the editor, it begins to suffer from this problem, and the original one no longer does.

I also recently encountered an error about my video card running out of texture memory, when juggling which scene to instance on play.  It goes away when I restart the editor.

Anyway, I will get on that demo project shortly.

**Comment by avencherus** [comment]
Reproducing it in a new project eludes me.  In my current project, I tried isolating it by painstakingly deleting segments one by one, and reproducing the stall by rerunning it.  Not any one single segment was to blame.  But the few instances I had total seemed to alleviate the lag.

I went through and removed all the scripts from all the objects contained in those scenes, but that didn't affect it noticeably.

I'm wondering now if maybe there is some junk data in some platform scene files, or if things get hectic when you put too many Sprites as children of an instance that is already child of an instance, and so on.

Going to take a deep breath, and try one more thing tonight, maybe see if one of the .tscn files has some kind of corruption by removing platforms bit by bit.

**Comment by avencherus** [comment]
Recreating it in an example project alludes me.  To find the source I painstakingly eliminated scenes from the main project, verifying the minutes of delay per test.  Not any single scene, segment, or script seemed to be affect it.  Overall reducing the scene count did reduce the delay.

So I don't know if this might be from nested scenes not liking having too many Sprites added to them as children, or some corrupt .tscn files.  If I find anything, or can reproduce it in an example project, I'll post back.

@reduz The main project is 52mb, if you've got the time to take a look at it, I can share it privately, but I don't know the best way to reach you.

**Comment by eon-s** [comment]
Can you check if there are textures loaded inside the scene files as text? (you will notice big text scene files, over 1mb), that usually slow down the scene loading when format is not binary.

**Comment by avencherus** [comment]
@eon-s 

They're about 3kb.  An example of one :

```
[gd_scene load_steps=5 format=1]

[ext_resource path="res://scripts/assets/platform.gd" type="Script" id=1]
[ext_resource path="res://resources/textures/levels/level01/atlases/big_platform.atex" type="Texture" id=2]
[ext_resource path="res://resources/textures/levels/level01/atlases/big_platform_top.atex" type="Texture" id=3]

[sub_resource type="ConvexPolygonShape2D" id=1]

custom_solver_bias = 0.0
points = Vector2Array( -444, 4, -418, 0, 400, 0, 426, 6, 350, 34, 0, 50, -374, 32 )

[node name="big_platform" type="StaticBody2D"]

input/pickable = false
shapes/0/shape = SubResource( 1 )
shapes/0/transform = Matrix32( 1, 0, 0, 1, 0, 0 )
shapes/0/trigger = false
collision/layers = 512
collision/mask = 0
constant_linear_velocity = Vector2( 0, 0 )
constant_angular_velocity = 0.0
friction = 1.0
bounce = 0.0
script/script = ExtResource( 1 )
__meta__ = {
"_edit_group_": true
}
highest_point = NodePath("respawn_points/point_1")
material_type = 0
light_mask = 17
can_pass_through = true
excludes_spawning = true
custom_slope = false
slope_max_angle = 50
slope_friction = 0.8
slope_accel = 1.0
shadow_can_rotate = true
color_override = Color( 1, 1, 1, 1 )

[node name="texture1" type="Sprite" parent="."]

visibility/light_mask = 49
transform/pos = Vector2( -4, 378 )
texture = ExtResource( 2 )

[node name="top" type="Sprite" parent="."]

visibility/light_mask = 49
transform/pos = Vector2( -7, -1 )
texture = ExtResource( 3 )

[node name="collider" type="CollisionPolygon2D" parent="."]

z/z = 2
build_mode = 0
polygon = Vector2Array( -444, 4, -418, 0, 400, 0, 426, 6, 350, 34, 0, 50, -374, 32 )
shape_range = Vector2( 0, 0 )
trigger = false

[node name="left_edge" type="Position2D" parent="."]

transform/pos = Vector2( -418, 0 )

[node name="right_edge" type="Position2D" parent="."]

transform/pos = Vector2( 400, 0 )

[node name="respawn_points" type="Node2D" parent="."]

[node name="point_1" type="Position2D" parent="respawn_points"]

transform/pos = Vector2( -350, 0 )

[node name="point_2" type="Position2D" parent="respawn_points"]

transform/pos = Vector2( -150, 0 )

[node name="point_3" type="Position2D" parent="respawn_points"]

[node name="point_4" type="Position2D" parent="respawn_points"]

transform/pos = Vector2( 150, 0 )

[node name="point_5" type="Position2D" parent="respawn_points"]

transform/pos = Vector2( 330, 0 )
```

**Comment by eon-s** [comment]
@avencherus you can write

<details>

`<details></details>`

</details>


for long blocks 
Look at the atex files too.

**Comment by avencherus** [comment]
@eon-s Sorry I don't follow.  What could be written, where, and what does it do?

I'll check and see if it might be the atlas textures as well.  It is sort of strange how bogged down my editor and game get even though it's only a few megabytes of textures over and over.

**Comment by reduz** [comment]
ah, I just realized this is godot 2.. I think this was optimized a lot in 3

**Comment by avencherus** [comment]
@reduz Oh, any advice on work-arounds to reduce development pain or what may be the culprit?  I've torn through as much as I could understand, and it seems like the editor is holding onto something in memory, because it will switch about depending on the last scene played.

I really hope this mostly editor only, I am a bit uneasy about publishing this game, I don't really understand what is or isn't happening on the backend when it comes to having so many nested scenes.  I'm going to eventually make as much as I can into placeholders, but I don't really know what else to do if it comes out unpleasantly laggy.

**Comment by eon-s** [comment]
@avencherus you can hide that big block of text inside a detail tag so all gets more readable.
atex can't be saved in another format (in case these result being the offending resources)?

**Comment by avencherus** [comment]
@eons-s Thanks.  I do think you're correct, this may be an atlas issue, I noticed my game was doing terribly in the build just recently.  What appears to have happened is a new sprite was added, I tossed into an existing atlas, and it appears to have caused it to upscale it into 4096 x 8192 texture.  (Some wonky placements.)

Once the atlas was split up some the game healed, but there still seems to be hang ups in the scene in the editor.  There were times the editor crashed with GPU memory errors, and those seem gone now.

When I finish some other work, I will try to optimize the atlases some more and see what happens.  I suspect you're right though.  2.1's editor is not so fond of big atlases.
